### PR TITLE
fix: Incorrect subscription name used by PulsarProvider when reader subscription name is provided already

### LINKED ISSUE #107: [BUG] Incorrect subscription name used by PulsarProvider when reader subscription name is provided already
**Describe the bug**
Consider a scenario:

`predefinedSubscription` option is not provided.
`pulsar.reader.subscriptionName` option is provided
Optionally - `subscriptionPrefix` may or maynot be provided

In this case, as per logic, pulsar-spark will go ahead and subscribe to the topic using subscription name derived from `$driverGroupIdPrefix-$topicPartition` ([here](https://github.com/streamnative/pulsar-spark/blob/master/src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala#L137)) which in turn is made up by `$subscriptionPrefix-$topicPartition` as that's what is passed to PulsarMetadataReader object [here](https://github.com/streamnative/pulsar-spark/blob/master/src/main/scala/org/apache/spark/sql/pulsar/PulsarProvider.scala#L69) and other places.

But the issue is that since the option `pulsar.reader.subscriptionName` is set, the actual reader created [here](https://github.com/streamnative/pulsar-spark/blob/master/src/main/scala/org/apache/spark/sql/pulsar/PulsarSourceRDD.scala#L65) will use a partition-agnostic common subscription name.

**To Reproduce**
mentioned above

**Expected behavior**
Ideally, if a low level (reader level) override is used, the PulsarProvider and PulsarMetadataReader should use that vallue instead of deriving it from subscriptionPrefix and topicPartition combination.

### PR DESCRIPTION
**Pull request by obobj**
Fixes #107 

### Motivation
Ideally, if a low level (reader level) override is used, the PulsarProvider and PulsarMetadataReader should use that vallue instead of deriving it from subscriptionPrefix and topicPartition combination.

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by nlu90** [comment]
@obobj 
Could you also update the doc accordingly?

File: src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala [hunk 1]
```diff
@@ -453,6 +454,7 @@ private[pulsar] case class PulsarMetadataReader(
             .topic(tp)
             .startMessageId(MessageId.earliest)
             .startMessageIdInclusive()
+            .loadConf(readerConfig)
```
Comment (nlu90) [inline src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala:457]: the `loadConf` method and other set methods should be mutually exclusive.

User either provides a `ReaderConfig` to load or set detailed configs with these setter method

**Review by nlu90 — CHANGES_REQUESTED** [review]

File: src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala (hunk 1 above)
Comment (obobj) [reply src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala]: Can we provide a priority to the user, the configuration based on the `pulsar.reader.` prefix will override the corresponding reader configuration

**Comment by nlu90** [comment]
@obobj This fix may not be needed. Please check my comment on the issue

**Comment by nlu90** [comment]
This PR is not needed.

**Pull request closed without merging**